apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2824,17 +2845,17 @@ func (u *UniterAPIv20) Watch(ctx context.Context, args params.Entities) (params.
 }
 
 // watchUnit returns a state notify watcher for the given unit.
-func (u *UniterAPI) watchUnit(tag names.Tag) (watcher.NotifyWatcher, error) {
-	entity0, err := u.st.FindEntity(tag)
+func (u *UniterAPI) watchUnit(ctx context.Context, tag names.UnitTag) (watcher.NotifyWatcher, error) {
+	unitName, err := coreunit.NewName(tag.Id())
 	if err != nil {
-		return nil, err
+		return nil, internalerrors.Errorf("parsing unit name: %w", err)
 	}
-	entity, ok := entity0.(state.NotifyWatcherFactory)
-	if !ok {
-		return nil, apiservererrors.NotSupportedError(tag, "watching")
+
+	watcher, err := u.applicationService.WatchUnitForLegacyUniter(ctx, unitName)
+	if err != nil {
```

Comment:
Shouldn't you check the `UnitNotFound` error and return a specific error?

Reply:
Yes good spot

Reply:
Resolved